language/borrow-graph/src/references/mod.rs

```diff
@@ -0,0 +1,138 @@
+    pub(crate) loc: Loc,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct BorrowEdges<Loc: Copy, Lbl: Clone + Ord>(
+    pub(crate) BTreeMap<RefID, BTreeSet<BorrowEdge<Loc, Lbl>>>,
+);
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct Ref<Loc: Copy, Lbl: Clone + Ord> {
+    // x is borrowed by y
+    pub(crate) borrowed_by: BorrowEdges<Loc, Lbl>,
+    // y borrows from x
+    pub(crate) borrows_from: BTreeSet<RefID>,
+    // true if mutable, false otherwise
+    pub(crate) mutable: bool,
+}
+
+//**************************************************************************************************
+// Impls
+//**************************************************************************************************
+
+impl<Loc: Copy, Lbl: Clone + Ord> BorrowEdge<Loc, Lbl> {
+    pub(crate) fn leq(&self, other: &Self) -> bool {
+        self == other || (!self.strong && paths::leq(&self.path, &other.path))
+    }
+}
+
+impl<Loc: Copy, Lbl: Clone + Ord> BorrowEdges<Loc, Lbl> {
+    pub(crate) fn new() -> Self {
+        Self(BTreeMap::new())
+    }
+}
+
+impl<Loc: Copy, Lbl: Clone + Ord> Ref<Loc, Lbl> {
+    pub(crate) fn new(mutable: bool) -> Self {
+        let borrowed_by = BorrowEdges::new();
+        let borrows_from = BTreeSet::new();
+        Self {
+            borrowed_by,
+            borrows_from,
+            mutable,
+        }
+    }
+}
+
+//**********************************************************************************************
+// Remap
+//**********************************************************************************************
+
+impl<Loc: Copy, Lbl: Clone + Ord> BorrowEdges<Loc, Lbl> {
+    pub(crate) fn remap_refs(&mut self, id_map: &BTreeMap<RefID, RefID>) {
+        for (old, new) in id_map {
+            if let Some(edges) = self.0.remove(old) {
+                self.0.insert(*new, edges);
+            }
+        }
+    }
+}
+
+impl<Loc: Copy, Lbl: Clone + Ord> Ref<Loc, Lbl> {
+    pub(crate) fn remap_refs(&mut self, id_map: &BTreeMap<RefID, RefID>) {
+        self.borrowed_by.remap_refs(id_map);
+        remap_set(&mut self.borrows_from, id_map)
+    }
+}
+
+//**********************************************************************************************
+// Traits
+//**********************************************************************************************
+
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+struct BorrowEdgeNoLoc<'a, Lbl: Clone> {
+    strong: bool,
+    path: &'a Path<Lbl>,
+}
+
+impl<'a, Lbl: Clone + Ord> BorrowEdgeNoLoc<'a, Lbl> {
+    fn new<Loc: Copy>(e: &'a BorrowEdge<Loc, Lbl>) -> Self {
+        BorrowEdgeNoLoc {
+            strong: e.strong,
+            path: &e.path,
+        }
+    }
+}
+
+impl<Loc: Copy, Lbl: Clone + Ord> PartialEq for BorrowEdge<Loc, Lbl> {
+    fn eq(&self, other: &BorrowEdge<Loc, Lbl>) -> bool {
+        BorrowEdgeNoLoc::new(self) == BorrowEdgeNoLoc::new(other)
+    }
+}
+
+impl<Loc: Copy, Lbl: Clone + Ord> Eq for BorrowEdge<Loc, Lbl> {}
```

Comment:
Taking this opportunity to learn Rusty matters.  What is the point of an empty impl declaration?

Reply:
I am getting lost in this BorrowEdgeNoLoc and Ord/PartialOrd business.  What is the objective here?

Reply:
> Taking this opportunity to learn Rusty matters. What is the point of an empty impl declaration?

This is how you implement `Eq`:
- First you implement `PartialEq`
- Then you implement `Eq` to tell it that the `PartialEq.eq` function is reflexive 

> I am getting lost in this BorrowEdgeNoLoc and Ord/PartialOrd business. What is the objective here?

I'll add a comment, its to skip the `Loc` fields